docs/source/daml/intro/3_Data.rst

```diff
@@ -0,0 +1,202 @@
+
+.. literalinclude:: daml/3_Data/Record.daml
+  :language: daml
+  :start-after: -- EQ_TEST_BEGIN
+  :end-before: -- EQ_TEST_END
+
+``Eq`` is a so-called type-class, which can be thought of like an interface in other languages. There are some other type-classes that can be derived automatically. Most prominently ``Show`` to get access to the function ``show``, equivalent to ``toString`` in many languages, and ``Ord``, which gives access to comparison operators ``<``, ``>``, ``<=``, ``>=``.
+
+It's a good idea to always derive ``Eq`` and ``Show`` using ``deriving (Eq, Show)``. The record types created using ``template T with`` do this automatically.
+
+Records can give the data on ``CashBalance`` a bit more structure:
+
+.. literalinclude:: daml/3_Data/Record.daml
+  :language: daml
+  :start-after: -- CASH_BALANCE_BEGIN
+  :end-before: -- CASH_BALANCE_END
+
+If you look at the resulting scenario view, you'll see that this still gives rise to one table. The records are expanded out into columns using dot notation.
+
+Variants
+~~~~~~~~
+
+Suppose now that you also wanted to keep track of cash in hand. Cash in hand doesn't have a bank, but you can't just leave ``bank`` empty. DAML doesn't have an equivalent to ``null``. Variants can express that cash can either be in hand or at a bank.
+
+.. literalinclude:: daml/3_Data/Variants.daml
+  :language: daml
+  :start-after: -- CASH_BALANCE_BEGIN
+  :end-before: -- CASH_BALANCE_END
+
+The way to read the declaration of ``Location`` is "*A Location either has value* ``InHand`` *OR has a value* ``InAccount a`` *where* ``a`` *is of type Account*". This is quite an explicit way to say that there may or may not be an ``Account`` associated with a ``CashBalance`` and gives both cases suggestive names. The generic way would be to use the built-in ``Optional`` type. The ``None`` value of type ``Optional a`` is the closest DAML has to a ``null`` value.
+
+.. literalinclude:: daml/3_Data/Variants.daml
+  :language: daml
+  :start-after: -- OPTIONAL_BEGIN
+  :end-before: -- OPTIONAL_END
+
+Variant types where none of the data constructors take a parameter are called enums:
+
+.. literalinclude:: daml/3_Data/Variants.daml
+  :language: daml
+  :start-after: -- ENUM_BEGIN
+  :end-before: -- ENUM_END
+
+Accessing data in variants requires distinguishing the different possible cases. It is no longer possible to access the account number of a ``Location`` directly, for example, as there may be no account number in case of ``InHand``. Instead, use *pattern matching* and either throw errors or return compatible types for all cases.
```

Comment:
Should we encourage people to reach for errors so soon in their DAML journey? I'm not very familiar with DAML (or Haskell for that matter), but in most languages I've worked with throwing exceptions is kind of a last resort and not particularly considered good style.

Reply:
I think so. ``assert`` and ``abort`` are part of the core DAML toolset needed to restrict your contract model. I see ``error`` as the ``abort`` for pure contexts. @neil-da , @martin-drhu-da  why do we even have both? ``error`` is perfectly usable in ``Action`` contexts, but not vice versa. Are there any circumstances where ``abort`` and ``error`` behave differently?